### PR TITLE
Label Rates UI Step

### LINKED ISSUE #506: [Labels] Rates step UI (one package only initially)
Pre-select whatever rate was chosen during checkout (if any) and allow the merchant to browse and choose a different rate. One package only for now.

### PR DESCRIPTION
**Pull request by jeffstieler**
Fixes #506.

Does not implement:
* Selecting a "similar" rate if customer-chosen rate isn't available.
* Price display options (no API equivalent discovered yet)

To test:
* Start "create label" flow
* If Origin and Destination are already OK, Rates step should show spinner upon opening modal
* Otherwise, once Origin and Destination are "green", Rates should fetch
* Change either Origin or Destination address, verify that Rates refresh
* Verify that when customer-chosen rate isn't available, notice is shown
  * Easiest way is to choose Priority X-Day shipping at checkout, and change one address to be farther away, such that it's now Priority Y-Day that's available

### CODE REVIEW

**Comment by kellychoffman** [comment]
One thing to add. I'll just quote myself from the Packages PR https://github.com/Automattic/woocommerce-connect-client/pull/531#issuecomment-245135478

> Lets add the confirmation button to this step :) I know it wasn't in the design, but when styling this I went through the flow about 20+ times. And when I get to this step after the address steps, I miss having a button that confirms the change and automatically collapses the card and brings you to the next step.

**Comment by allendav** [comment]
Needs a rebase :)

**Comment by allendav** [comment]
In this situation, we should disable the Buy and Print button:

screen shot 2016-09-07 at 3 45 44 pm

**Comment by allendav** [comment]
Let's sort the rates by price

screen shot 2016-09-07 at 4 01 17 pm

**Comment by allendav** [comment]
For a single package, display the selected rate in the collapsed expander's header

screen shot 2016-09-07 at 4 03 01 pm

**Comment by jeffstieler** [comment]
> Let's sort the rates by price

Can we make that the server's job?

> For a single package, display the selected rate in the collapsed expander's header

Fixed in https://github.com/Automattic/woocommerce-connect-client/pull/533/commits/98bd322fb8c8f5f25e966173b9982b6f4e8979b6.

> In this situation, we should disable the Buy and Print button:

Fixed by side effect in https://github.com/Automattic/woocommerce-connect-client/pull/533/commits/1424f4c1858f64162d7909b148edd9bf22d937ee.

File: classes/class-wc-connect-shipping-label.php [hunk 1]
```diff
@@ -106,36 +115,61 @@ protected function get_selected_rates( WC_Order $order ) {
 			$packages = json_decode( $shipping_method[ 'wc_connect_packages' ], true );
 			$rates = array();
 
-			foreach( $packages as $package ) {
+			foreach( $packages as $idx => $package ) {
 				if ( ! $package[ 'service_id' ] ) {
 					return array();
 				}
-				$rates[] = $package[ 'service_id' ];
+				$id = isset( $package[ 'id' ] ) ? $package[ 'id' ] : "box_{$idx}";
```
Comment (allendav) [inline classes/class-wc-connect-shipping-label.php]: Would you add a brief comment explaining the need for this?

File: classes/class-wc-connect-shipping-label.php [hunk 2]
```diff
@@ -106,36 +115,61 @@ protected function get_selected_rates( WC_Order $order ) {
...
+				$rates[ $id ] = $package[ 'service_id' ];
 			}
 
 			return $rates;
 		}
 
-		protected function get_form_data( WC_Order $order ) {
-			$form_data = array();
+		protected function format_address_for_server( $address ) {
```
Comment (allendav) [inline classes/class-wc-connect-shipping-label.php]: "server" could be ambiguous to someone unfamiliar with the code. Would you consider "format_address_for_client_api" or something like that instead?

File: client/shipping-label/views/index.js [hunk 3]
```diff
@@ -42,6 +62,7 @@ function mapStateToProps( state, { storeOptions } ) {
 	return {
 		shippingLabel: state.shippingLabel,
 		errors: getFormErrors( state, storeOptions ),
+		ratesTotal: getRatesTotal( state.shippingLabel.form.rates.values, state.shippingLabel.form.rates.available ),
```
Comment (allendav) [inline client/shipping-label/views/index.js]: I noticed you didn't add customerRateChoices to propTypes.  Would you add that?

File: client/shipping-label/views/steps/rates/index.js [hunk 4]
```diff
@@ -26,11 +110,11 @@ const RatesStep = ( { form, values, available, storeOptions, labelActions, error
...
 RatesStep.propTypes = {
 	form: PropTypes.object.isRequired,
-	values: PropTypes.array.isRequired,
-	available: PropTypes.array.isRequired,
+	values: PropTypes.object.isRequired,
+	available: PropTypes.object.isRequired,
 	labelActions: PropTypes.object.isRequired,
 	storeOptions: PropTypes.object.isRequired,
-	errors: PropTypes.array.isRequired,
+	errors: PropTypes.object.isRequired,
```
Comment (allendav) [inline client/shipping-label/views/steps/rates/index.js]: Please add customerRateChoices

File: client/shipping-label/views/steps/rates/list.js [hunk 5]
```diff
@@ -1,61 +1,80 @@
...
-
+		const selectedRate = selectedRates[ pckg.id ] || '';
+		const packageRates = get( availableRates, [ pckg.id, 'rates' ], [] );
 		const valuesMap = { '': __( 'Select one...' ) };
-		Object.keys( availableRates[ index ] ).forEach( ( serviceId ) => {
-			const rateObject = availableRates[ index ][ serviceId ];
-			valuesMap[ serviceId ] = rateObject.name + ' (' + currencySymbol + rateObject.rate.toFixed( 2 ) + ')';
+
+		packageRates.forEach( ( rateObject ) => {
+			valuesMap[ rateObject.service_id ] = rateObject.title + ' (' + currencySymbol + rateObject.rate.toFixed( 2 ) + ')';
```
Comment (allendav) [inline client/shipping-label/views/steps/rates/list.js]: Seems like our endpoints (on the merchant host and on our server) should give prices to two decimal points and we shouldn't need to do things like toFixed.  Do we really need this?  If so, let's add issues to fix it at the endpoints.

File: client/shipping-label/views/steps/rates/list.js [hunk 6]
```diff
@@ -1,61 +1,80 @@
...
-		</ul>
+		</div>
 	);
 };
 
 ShippingRates.propTypes = {
 	id: PropTypes.string.isRequired,
-	selectedRates: PropTypes.array.isRequired,
-	availableRates: PropTypes.array.isRequired,
+	selectedRates: PropTypes.object.isRequired,
```
Comment (allendav) [inline client/shipping-label/views/steps/rates/list.js]: This is whatever the merchant has selected in the dialog, right?  And not necessarily what the customer selected.  Let's add a comment maybe clarifying that for developers new to the code.

**Comment by allendav** [comment]
>> Let's sort the rates by price
> Can we make that the server's job?
Sure - please open an issue

**Comment by kellychoffman** [comment]
This might be a separate issue, but this happens when I step through the entire flow: I get to packages, click the confirm button 'use this package', the packages step collapses, but the rates step doesn't open.

File: client/shipping-label/views/steps/rates/list.js [hunk 7]
```diff
@@ -1,61 +1,80 @@
...
+import Notice from 'components/notice';
 import { translate as __ } from 'lib/mixins/i18n';
+import { sprintf } from 'sprintf-js';
+import get from 'lodash/get';
+
+const renderRateNotice = ( show ) => {
+	if ( show ) {
+		return (
+			<Notice
+				className="rate-not-available"
```
Comment (kellychoffman) [inline client/shipping-label/views/steps/rates/list.js]: Fixed in cc4c529

**Comment by jeffstieler** [comment]
>>> Let's sort the rates by price
>> Can we make that the server's job?
> Sure - please open an issue

https://github.com/Automattic/woocommerce-connect-server/issues/462

File: client/shipping-label/views/index.js (hunk 3 above)
Comment (jeffstieler) [inline client/shipping-label/views/index.js]: Added in https://github.com/Automattic/woocommerce-connect-client/pull/533/commits/8170c7f2a15a3f9542d4aad4e44b69526c5293ea.

File: client/shipping-label/views/steps/rates/index.js (hunk 4 above)
Comment (jeffstieler) [inline client/shipping-label/views/steps/rates/index.js]: Added in https://github.com/Automattic/woocommerce-connect-client/pull/533/commits/8a55916c833e85b9d4969775df397441ff772645.

File: client/shipping-label/views/steps/rates/list.js (hunk 5 above)
Comment (DanReyLop) [inline client/shipping-label/views/steps/rates/list.js]: `toFixed` means that the float `9.1` will be displayed as `"9.10"`, which is what we want for displaying money amounts.

File: client/shipping-label/views/dialog.js [hunk 8]
```diff
@@ -2,90 +2,116 @@ import React from 'react';
...
+	maybeGetNewRates( nextProps ) {
+		const useNormalizedOrigin = nextProps.form.origin.selectNormalized;
+		const useNormalizedDestination = nextProps.form.destination.selectNormalized;
 
-	const getTotalCost = () => {
-		const ratesInfo = props.form.rates;
-		const ratesCost = ratesInfo.values.map( ( rateId, index ) => ratesInfo.available[ index ][ rateId ].rate );
-		return sum( ratesCost ).toFixed( 2 );
-	};
+		// Determine if we need to compare user-entered values, or normalized ones
```
Comment (DanReyLop) [inline client/shipping-label/views/dialog.js]: This component is "too smart". It's better to have as much logic as possible in one place, and in this case the place should be `actions.js`. The `props` you are checking are part of the `state`, so they are available for the `actions`.

A good way of knowing that something shouldn't be in the component, is that you needed to change from a functional component to a Class component just for this logic.

File: client/shipping-label/views/steps/rates/list.js (hunk 6 above)
Comment (jeffstieler) [inline client/shipping-label/views/steps/rates/list.js]: What about renaming this to `values` like it is in state?

File: client/shipping-label/views/dialog.js (hunk 8 above)
Comment (jeffstieler) [inline client/shipping-label/views/dialog.js]: Agreed - fixed in https://github.com/Automattic/woocommerce-connect-client/pull/533/commits/c902be65c71b39bc216404ca6604707eca80fec1.

**Comment by jeffstieler** [comment]
> This might be a separate issue, but this happens when I step through the entire flow: I get to packages, click the confirm 'use this package' button, the packages step collapses, but the rates step doesn't open.

@kellychoffman I've changed the behavior to close the packages step and kick off rates retrieval when "use this package" is clicked.

File: client/shipping-label/views/steps/rates/list.js (hunk 5 above)
Comment (jeffstieler) [inline client/shipping-label/views/steps/rates/list.js]: Created an issue here: https://github.com/Automattic/woocommerce-connect-server/issues/465

**Comment by allendav** [comment]
I get this when I attempt to buy a label after selecting a rate:

```
[08-Sep-2016 20:33:49 UTC] PHP Fatal error:  Uncaught Error: Call to a member function get_selected_payment_method_id() on null in /Applications/MAMP/htdocs/wp-content/plugins/woocommerce-connect-client/classes/class-wc-rest-connect-shipping-label-controller.php:52
Stack trace:
#0 /Applications/MAMP/htdocs/wp-includes/rest-api/class-wp-rest-server.php(907): WC_REST_Connect_Shipping_Label_Controller->update_items(Object(WP_REST_Request))
#1 /Applications/MAMP/htdocs/wp-includes/rest-api/class-wp-rest-server.php(326): WP_REST_Server->dispatch(Object(WP_REST_Request))
#2 /Applications/MAMP/htdocs/wp-includes/rest-api.php(147): WP_REST_Server->serve_request('/wc/v1/connect/...')
#3 /Applications/MAMP/htdocs/wp-includes/plugin.php(600): rest_api_loaded(Object(WP))
#4 /Applications/MAMP/htdocs/wp-includes/class-wp.php(386): do_action_ref_array('parse_request', Array)
#5 /Applications/MAMP/htdocs/wp-includes/class-wp.php(725): WP->parse_request('')
#6 /Applications/MAMP/htdocs/wp-includes/functions.php(963): WP->main('')
#7 /Applications/MAMP/htdocs/wp-blog-he in /Applications/MAMP/htdocs/wp-content/plugins/woocommerce-connect-client/classes/class-wc-rest-connect-shipping-label-controller.php on line 52
```

**Comment by allendav** [comment]
@jeffstieler - let's get rid of the "The service and rate chosen..." notice entirely for now, open a separate issue for hooking it up, and proceed with this.

**Comment by jeffstieler** [comment]
> let's get rid of the "The service and rate chosen..." notice entirely for now, open a separate issue for hooking it up, and proceed with this.

https://github.com/Automattic/woocommerce-connect-client/issues/546

**Comment by allendav** [comment]
:shipit:

**Pull request merged**